File: backend/app/services/replenishment_service.py
from __future__ import annotations

from datetime import datetime, timedelta, timezone
from decimal import Decimal

from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.errors import AppError
from app.models import (
    PurchaseMaterial,
    StockMaterial,
    StockReplenishmentPolicy,
)
from app.schemas import (
    PurchaseMaterialCreate,
    ReplenishmentDefaultsRead,
    ReplenishmentDraftCreate,
    ReplenishmentDraftRead,
    ReplenishmentPolicyWrite,
)
from app.services.common import validate_quantity_precision, validate_version
from app.services.inventory_service import recent_outbound_consumption
from app.services.material_service import create_purchase_material, get_stock_material

ZERO = Decimal("0")
SHANGHAI = timezone(timedelta(hours=8))


async def replenishment_defaults(session: AsyncSession) -> ReplenishmentDefaultsRead:
    latest_responsible = await session.scalar(
        select(PurchaseMaterial.purchase_responsible)
        .where(
            func.trim(PurchaseMaterial.purchase_responsible).not_in(("", "\\", "/", "—", "-"))
        )
        .order_by(PurchaseMaterial.id.desc())
        .limit(1)
    )
    return ReplenishmentDefaultsRead(
        purchase_responsible=latest_responsible or "",
        demand_date=datetime.now(SHANGHAI).date(),
    )


async def set_policy(
    session: AsyncSession,
    material: StockMaterial,
    data: ReplenishmentPolicyWrite,
) -> StockMaterial:
    validate_quantity_precision(data.minimum_qty, material.unit.decimal_places)
    policy = material.replenishment_policy
    if policy is None:
        policy = StockReplenishmentPolicy(
            stock_material_id=material.id,
            minimum_qty=data.minimum_qty,
            enabled=data.enabled,
        )
        session.add(policy)
        material.replenishment_policy = policy
    else:
        validate_version(data.version, policy.version)
        policy.minimum_qty = data.minimum_qty
        policy.enabled = data.enabled
        policy.version += 1
    await session.flush()
    return material


async def create_replenishment_draft(
    session: AsyncSession,
    stock_material_id: int,
    data: ReplenishmentDraftCreate,
) -> ReplenishmentDraftRead:
    stock = await get_stock_material(session, stock_material_id)
    if stock.balance is None:
        raise AppError("BALANCE_MISSING", "库存余额记录不存在", status_code=409)
    policy = stock.replenishment_policy
    if policy is None or not policy.enabled or stock.balance.quantity > policy.minimum_qty:
        raise AppError("NOT_LOW_STOCK", "该物资当前不在低库存范围", status_code=409)

    suggested = (await recent_outbound_consumption(session, [stock.id])).get(stock.id, ZERO)
    validate_quantity_precision(data.planned_qty, stock.unit.decimal_places)

    previous_code = await session.scalar(
        select(PurchaseMaterial.material_code)
        .where(
            PurchaseMaterial.stock_material_id == stock.id,
            PurchaseMaterial.material_code.is_not(None),
        )
        .order_by(PurchaseMaterial.id.desc())
        .limit(1)
    )
    quantity_note = f"补库计划：建议申购 {suggested} {stock.unit.name}"
    if data.planned_qty != suggested:
        quantity_note += f"，确认计划 {data.planned_qty} {stock.unit.name}"
    note_parts = [part for part in [stock.remark, quantity_note] if part]
    purchase = await create_purchase_material(
        session,
        PurchaseMaterialCreate(
            plan_date=data.demand_date,
            material_code=previous_code,
            name=stock.name,
            model_spec=stock.model_spec,
            unit_id=stock.unit_id,
            actual_demand_person=data.actual_demand_person,
            purchase_responsible=data.purchase_responsible,
            planned_qty=data.planned_qty,
            usage="低库存补库",
            remark="；".join(note_parts),
            stock_material_id=stock.id,
            image_ids=[link.file_id for link in stock.images],
        ),
    )
    return ReplenishmentDraftRead(next="purchase_material", resource_id=purchase.id)

File: server/internal/service/replenishment_service.go
package service

import (
	"net/http"
	"time"

	"github.com/shopspring/decimal"
	"gorm.io/gorm"

	apperrors "github.com/yangrucheng/materials-manager/server/internal/errors"
	"github.com/yangrucheng/materials-manager/server/internal/models"
)

// ShanghaiTZ 上海时区（业务时间基准）。
var ShanghaiTZ = time.FixedZone("SHANGHAI", 8*3600)

// ReplenishmentDefaults 返回补库默认值。
func ReplenishmentDefaults(db *gorm.DB) (responsible string, demandDate time.Time, appErr *apperrors.AppError) {
	var latest string
	err := db.Model(&models.PurchaseMaterial{}).
		Where("TRIM(purchase_responsible) NOT IN ?", []string{"", "\\", "/", "—", "-"}).
		Order("id DESC").Limit(1).Pluck("purchase_responsible", &latest).Error
	if err != nil {
		return "", time.Time{}, DatabaseError(err)
	}
	return latest, time.Now().In(ShanghaiTZ), nil
}

// SetReplenishmentPolicy 保存安全库存策略（乐观锁）。
func SetReplenishmentPolicy(db *gorm.DB, material *models.StockMaterial, minimumQty decimal.Decimal, enabled bool, version int) *apperrors.AppError {
	if appErr := ValidateQuantityPrecision(minimumQty); appErr != nil {
		return appErr
	}
	var policy models.StockReplenishmentPolicy
	err := db.Where("stock_material_id = ?", material.ID).First(&policy).Error
	if IsNotFound(err) {
		policy = models.StockReplenishmentPolicy{
			StockMaterialID: material.ID,
			MinimumQty:      models.Decimal{Decimal: minimumQty},
			Enabled:         enabled,
			CreatedAt:       models.UTCNow(),
			UpdatedAt:       models.UTCNow(),
			Version:         1,
		}
		if err := db.Create(&policy).Error; err != nil {
			return DatabaseError(err)
		}
		return nil
	}
	if err != nil {
		return DatabaseError(err)
	}
	if appErr := ValidateVersion(version, policy.Version); appErr != nil {
		return appErr
	}
	policy.MinimumQty = models.Decimal{Decimal: minimumQty}
	policy.Enabled = enabled
	policy.Version++
	policy.UpdatedAt = models.UTCNow()
	if err := db.Model(&models.StockReplenishmentPolicy{}).Where("stock_material_id = ?", material.ID).
		Updates(map[string]any{
			"minimum_qty": minimumQty.String(),
			"enabled":     enabled,
			"version":     gorm.Expr("version + 1"),
			"updated_at":  policy.UpdatedAt,
		}).Error; err != nil {
		return DatabaseError(err)
	}
	return nil
}

// CreateReplenishmentDraft 从低库存一键生成补库申购计划草稿。
func CreateReplenishmentDraft(db *gorm.DB, stockMaterialID int64, demandDate time.Time, actualDemandPerson, purchaseResponsible string, plannedQty decimal.Decimal) (*models.PurchaseMaterial, *apperrors.AppError) {
	stock, appErr := GetStockMaterial(db, stockMaterialID)
	if appErr != nil {
		return nil, appErr
	}
	var balance models.StockBalance
	err := db.Where("stock_material_id = ?", stockMaterialID).First(&balance).Error
	if IsNotFound(err) {
		return nil, apperrors.New("BALANCE_MISSING", "库存余额记录不存在", http.StatusConflict, nil)
	}
	if err != nil {
		return nil, DatabaseError(err)
	}
	var policy models.StockReplenishmentPolicy
	policyErr := db.Where("stock_material_id = ?", stockMaterialID).First(&policy).Error
	policyMissing := IsNotFound(policyErr)
	if policyErr != nil && !policyMissing {
		return nil, DatabaseError(policyErr)
	}
	if policyMissing || !policy.Enabled || balance.Quantity.Decimal.GreaterThan(policy.MinimumQty.Decimal) {
		return nil, apperrors.New("NOT_LOW_STOCK", "该物资当前不在低库存范围", http.StatusConflict, nil)
	}
	consumption, err := RecentOutboundConsumption(db, []int64{stock.ID}, time.Now().UTC())
	if err != nil {
		return nil, DatabaseError(err)
	}
	suggested := consumption[stock.ID]
	if appErr := ValidateQuantityPrecision(plannedQty); appErr != nil {
		return nil, appErr
	}
	var previousCode string
	_ = db.Model(&models.PurchaseMaterial{}).
		Where("stock_material_id = ? AND material_code IS NOT NULL", stock.ID).
		Order("id DESC").Limit(1).Pluck("material_code", &previousCode).Error

	unit := stock.UnitName
	quantityNote := "补库计划：建议申购 " + serializeDecimal(suggested) + " " + unit
	if !plannedQty.Equal(suggested) {
		quantityNote += "，确认计划 " + serializeDecimal(plannedQty) + " " + unit
	}
	var noteParts []string
	if stock.Remark != nil && *stock.Remark != "" {
		noteParts = append(noteParts, *stock.Remark)
	}
	noteParts = append(noteParts, quantityNote)

	_, images, appErr := StockMaterialImages(db, stock.ID)
	if appErr != nil {
		return nil, appErr
	}
	imageIDs := make([]string, 0, len(images))
	for _, f := range images {
		imageIDs = append(imageIDs, f.ID)
	}
	var codePtr *string
	if previousCode != "" {
		codePtr = &previousCode
	}
	var stockIDPtr *int64
	stockID := stock.ID
	stockIDPtr = &stockID

	purchase, appErr := CreatePurchaseMaterial(db,
		demandDate,
		ptrStr(codePtr), "", "正常", "HXNI 检修维护部",
		stock.Name, stock.ModelSpec, stock.UnitName,
		actualDemandPerson, purchaseResponsible,
		plannedQty,
		"低库存补库", "", joinStrings(noteParts, "；"),
		stockIDPtr, "NORMAL", imageIDs)
	if appErr != nil {
		return nil, appErr
	}
	return purchase, nil
}

func ptrStr(s *string) string {
	if s == nil {
		return ""
	}
	return *s
}

func joinStrings(parts []string, sep string) string {
	out := ""
	for i, p := range parts {
		if i > 0 {
			out += sep
		}
		out += p
	}
	return out
}
